core/src/main/scala/org/apache/spark/status/AppStatusStore.scala

```diff
@@ -733,6 +734,15 @@ private[spark] class AppStatusStore(
 
   def close(): Unit = {
     store.close()
+    cleanUpStorePath()
+  }
+
+  private def cleanUpStorePath(): Unit = {
+    storePath.foreach { p =>
+      if (p.exists()) {
+        p.listFiles().foreach(Utils.deleteRecursively)
+      }
+    }
```

Comment:
Anything under the application directory will be cleaned up, right ?
Are we expecting this to be set outside of it ? (If yes, we will have dangling files any abnormal termination).

Btw, why not simply `storePath.foreach(Utils.deleteRecursively)

Reply:
[c1b5447](https://github.com/apache/spark/pull/39226/commits/c1b54476bbad9f768588d23e12083a7880fec5f0) change to use `storePath.foreach(Utils.deleteRecursively)`

Reply:
Yes, abnormal termination will generate dangling files,  users need to manually cleanup in this scenario.

There's another problem I need to mention:

- Start spark-shell twice in the same directory with `spark.ui.store.path` configured:
   - The first app start successfully and use `spark.ui.store.path` to store LiveUI data
   - The second app will also start successfully, but `InMemoryStore` will be used to store Live UI data due to `org.rocksdb.RocksDBException: While lock file: /${baseDir}/listing.rdb/LOCK: Resource temporarily unavailable`


Should we solve this problem, such as creating a subdirectory under `spark.ui.store.path` for each different App to ensure the isolation of rocksdb usage? (However, if we do this, do we still need to add the `cleanup` config? The app will use different directories after restart)

Reply:
There are three separate cases here, though related, let us look at them individually.

a) Should we cleanup ?

For cluster mode, it is not very useful - since the cluster manager will do the same.

In client mode, like the excellent `spark-shell` example you gave @LuciferYang, this is indeed very relevant.
Given this, let us keep both modes consistent and cleanup for both (like what we do for block manager dirs for example).


b) Should we need a config to clean up live ui local files.

I would argue always clean up - we dont need config for it. Just as we cleanup all other temp paths created during spark lifecycle.
Is there any case where we do need to control this ? +CC @gengliangwang 


c) Issue with rocks db locking.

We should treat this as a bug - couple of ways to fix it, though I would say we can simply follow what we already have in spark - for example, use `getLocalDir`.
Is there a reason to expose the ability to configure this path ? Or should it simply be an implementation detail, and something spark automatically creates during start and remove at stop ?

Reply:
Got it, I think this is a good suggestion and agree with you, let's wait for @gengliangwang

Reply:
> I would argue always clean up - we dont need config for it. Just as we cleanup all other temp paths created during spark lifecycle.

OK let's simply cleanup them without a configuration now. My initial concern is that some users may want to leverage the files in SHS. Since we don't have a design for it, we can just remove them.

Reply:
> Is there a reason to expose the ability to configure this path ?

Do you mean the RocksDB file path? The size can be big if a user increases the retention limits. Thus a user should be able to choose the file path for it, instead of automatically choosing a temp dir. 
If there is a corrupted RocksDB file during startup, shall we just remove them and try creating the RocksDB file again? There is such code path: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/KVUtils.scala#L155

Reply:
> There's another problem I need to mention:
> 
> * Start spark-shell twice in the same directory with `spark.ui.store.path` configured:
>   
>   * The first app start successfully and use `spark.ui.store.path` to store LiveUI data
>   * The second app will also start successfully, but `InMemoryStore` will be used to store Live UI data due to
> 
> `org.rocksdb.RocksDBException: While lock file: /${baseDir}/listing.rdb/LOCK: Resource temporarily unavailable`

@gengliangwang This scenario is a special case, we should not delete it,  otherwise, the first app will be affected(in the scenario, the first app not stop)

Reply:
I think we can let the user specify a basedir and create a subdirectory under the basedir to ensure that rocksdb is isolated with app granularity. At the same time, this subdirectory is automatically created when Spark starts and automatically deleted when it stops.

I want to change this pr according to this ideas, and then we can see if it is acceptable @gengliangwang @mridulm

Reply:
A little busy today, will update later

Reply:
I was specifically referring to the context of driver using the kvstore for the live UI data.
The codepath you [referred to](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/KVUtils.scala#L155) is for use in history server (particularly after a restart/upgrade of history server) - not relevant for driver.
It becomes relevant to driver when multiple applications end up using the same path - primarily in client mode.

Reply:
Let me think more

Reply:
The current code uses a strategy similar to `DiskBlockManager` to generate a temporary directory for RocksDB (without exposing `spark.ui.store.path` to users).

However, I think it is reasonable to expose `spark.ui.store.path` to users: when multiple long running Spark Apps(client mode) are running on same machine  and RockDB is used to store the Live UI, the default use of `/tmp`  to store data may cause `no space`, because `/tmp` directory is generally limited (about 5~7G space)

WDYT @gengliangwang @mridulm ? If `spark.ui.store.path` needs to be exposed to users, I need to revert the last commit.